static/app/utils/profiling/flamegraph.ts

```diff
@@ -0,0 +1,212 @@
+import {lastOfArray} from 'sentry/utils';
+
+import {Rect} from './gl/utils';
+import {Profile} from './profile/profile';
+import {CallTreeNode} from './callTreeNode';
+import {FlamegraphFrame} from './flamegraphFrame';
+
+export class Flamegraph {
+  profile: Profile;
+  frames: FlamegraphFrame[] = [];
+
+  name: string;
+  profileIndex: number;
+  startedAt: number;
+  endedAt: number;
+
+  inverted?: boolean = false;
+  leftHeavy?: boolean = false;
+
+  colors: Map<string | number, number[]> | null = null;
+
+  depth = 0;
+  duration = 0;
+  configSpace: Rect = new Rect(0, 0, 0, 0);
+
+  constructor(
+    profile: Profile,
+    profileIndex: number,
+    inverted = false,
+    leftHeavy = false
+  ) {
+    this.inverted = inverted;
+    this.leftHeavy = leftHeavy;
+
+    // @TODO check if we can not keep a reference to the profile
+    this.profile = profile;
+
+    this.duration = profile.duration;
+    this.profileIndex = profileIndex;
+    this.name = profile.name;
+
+    this.startedAt = profile.startedAt;
+    this.endedAt = profile.endedAt;
+
+    this.frames = leftHeavy
+      ? this.buildLeftHeavyGraph(profile)
+      : this.buildCallOrderGraph(profile);
+
+    if (this.frames.length) {
+      this.configSpace = new Rect(0, 0, this.duration, this.depth);
+    } else {
+      // If we have no frames, set the trace duration to 1 second so that we can render a placeholder grid
+      this.configSpace = new Rect(0, 0, 1_000_000, 0);
```

Comment:
Are the units always going to be `microseconds`? I see that it's possible to specify a unit in `Profile`. Should that be used to compute this value?

Reply:
Ah correct, we need to set that too, I missed that 🙏🏼